### DISCUSSION TITLE
Ancestry query

### DISCUSSION THREAD

**Discussion by jedi1992** [EvgSkv/logica#110] [Q&A] [answered]
( Looks like Github does not have a delete discussion option. Realized the answer after posting the Question. So converting it to  a self Q&A) 

 Hi
In the context of the Queen Victoria example is this tweet correct https://twitter.com/EvgSkvDev/status/1395988919582265350
Or
should the query be like this
AncestryInDownDirection(a,b) distinct :-
              Parent(a,b) | AncestryInDownDirection(a,x), Parent(x,b) ;

Just realized that I missed a bracket after the disjunction |.
Interestingly the above  query works both ways with and without bracket.

But the one in the tweet would work only with the bracket.   

Thanks

**Comment by EvgSkv** [comment, marked as answer]
That's right, this form works as well. If you don't put any parenthesis then it's interpreted as
```
AncestryInDownDirection(a,b) distinct :-
  (Parent(a,b) | AncestryInDownDirection(a,x)), Parent(x,b) ;
```
Which is still a correct definition of ancestry. Even though maybe not very intuitive :-)
It's more intuitive with parenthesis
```
AncestryInDownDirection(a,b) distinct :-
  Parent(a,b) | (AncestryInDownDirection(a,x), Parent(x,b)) ;
```

Note that the depth of ancestry as a function of the number of recursive steps grows linearly in this rule and exponentially in the rule from the tweet. So if your tree is deep, but is manageable then the query from the tweet is preferred. On the other hand if you want only a small limited depth of ancestry then the rule as defined above as preferred.
